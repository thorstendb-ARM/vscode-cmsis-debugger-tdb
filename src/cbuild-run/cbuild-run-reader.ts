/**
 * Copyright 2025 Arm Limited
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as yaml from 'yaml';
import { CbuildRunRootType, CbuildRunType } from './cbuild-run-types';
import { FileReader, VscodeFileReader } from '../desktop/file-reader';
import { getCmsisPackRootPath } from '../utils';

const CMSIS_PACK_ROOT_ENVVAR = '${CMSIS_PACK_ROOT}';

export class CbuildRunReader {
    private cbuildRun: CbuildRunType | undefined;

    constructor(private reader: FileReader = new VscodeFileReader()) {}

    public hasContents(): boolean {
        return !!this.cbuildRun;
    }

    public getContents(): CbuildRunType | undefined {
        return this.cbuildRun;
    }

    public async parse(filePath: string): Promise<void> {
        const fileContents = await this.reader.readFileToString(filePath);
        const fileRoot = yaml.parse(fileContents) as CbuildRunRootType;
        this.cbuildRun = fileRoot ? fileRoot['cbuild-run'] : undefined;
        if (!this.cbuildRun) {
            throw new Error(`Invalid '*.cbuild-run.yml' file: ${filePath}`);
        }
    }

    public getSvdFilePaths(cmsisPackRoot?: string, pname?: string): string[] {
        if (!this.cbuildRun) {
            return [];
        }
        // Get SVD file descriptors
        const systemDescriptions = this.cbuildRun['system-descriptions'];
        const svdFileDescriptors = systemDescriptions?.filter(descriptor => descriptor.type === 'svd') ?? [];
        if (svdFileDescriptors.length === 0) {
            return [];
        }
        // Replace potential ${CMSIS_PACK_ROOT} placeholder
        const effectiveCmsisPackRoot = cmsisPackRoot ?? getCmsisPackRootPath();
        // Map to copies, leave originals untouched
        const filteredSvdDescriptors = pname ? svdFileDescriptors.filter(descriptor => descriptor.pname === pname): svdFileDescriptors;
        const svdFilePaths = filteredSvdDescriptors.map(descriptor => `${effectiveCmsisPackRoot
            ? descriptor.file.replaceAll(CMSIS_PACK_ROOT_ENVVAR, effectiveCmsisPackRoot)
            : descriptor.file}`);
        return svdFilePaths;
    }

    public getPnames(): string[] {
        if (!this.cbuildRun) {
            return [];
        }
        const processors = this.cbuildRun['debug-topology']?.processors;
        const pnameProcessors = processors?.filter(p => p.pname);
        if (!pnameProcessors?.length) {
            return [];
        }
        return pnameProcessors.map(p => p.pname!);
    }

}
