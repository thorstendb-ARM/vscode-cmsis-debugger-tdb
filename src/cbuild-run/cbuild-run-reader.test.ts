/**
 * Copyright 2025 Arm Limited
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { CbuildRunReader } from './cbuild-run-reader';

const TEST_CBUILD_RUN_FILE = 'test-data/multi-core.cbuild-run.yml'; // Relative to repo root
const TEST_FILE_PATH = 'test-data/fileReaderTest.txt'; // Relative to repo root

describe('CbuildRunReader', () => {

    describe('Parser', () => {
        it('successfully parses a *.cbuild-run.yml file', async () => {
            const cbuildRunReader = new CbuildRunReader();
            await expect(cbuildRunReader.parse(TEST_CBUILD_RUN_FILE)).resolves.not.toThrow();
            expect(cbuildRunReader.hasContents()).toBe(true);
            const contents = cbuildRunReader.getContents();
            expect(contents).toBeDefined();
            expect(contents).toMatchSnapshot();
        });

        it('throws if it parses something other than a *.cbuild-run.yml file and correctly responds to raw contents calls', async () => {
            const expectedError = /Invalid '\*\.cbuild-run\.yml' file: .*test-data\/fileReaderTest\.txt/;
            const cbuildRunReader = new CbuildRunReader();
            await expect(cbuildRunReader.parse(TEST_FILE_PATH)).rejects.toThrow(expectedError);
            expect(cbuildRunReader.hasContents()).toBe(false);
            expect(cbuildRunReader.getContents()).toBeUndefined();
        });

        it('correctly responds to raw contents calls if nothing is parsed', () => {
            const cbuildRunReader = new CbuildRunReader();
            expect(cbuildRunReader.hasContents()).toBe(false);
            expect(cbuildRunReader.getContents()).toBeUndefined();
        });
    });

    describe('Extract Values', () => {
        let cbuildRunReader: CbuildRunReader;

        beforeEach(() => {
            cbuildRunReader = new CbuildRunReader();
        });

        it.each([
            {
                info: 'no pname',
                pname: undefined,
                expectedSvdPaths: [
                    '/my/pack/root/MyVendor/MyDevice/1.0.0/Debug/SVD/MyDevice_Core0.svd',
                    '/my/pack/root/MyVendor/MyDevice/1.0.0/Debug/SVD/MyDevice_Core1.svd'
                ]
            },
            {
                info: 'Core0',
                pname: 'Core0',
                expectedSvdPaths: [
                    '/my/pack/root/MyVendor/MyDevice/1.0.0/Debug/SVD/MyDevice_Core0.svd',
                ]
            },
            {
                info: 'Core1',
                pname: 'Core1',
                expectedSvdPaths: [
                    '/my/pack/root/MyVendor/MyDevice/1.0.0/Debug/SVD/MyDevice_Core1.svd'
                ]
            },
        ])('returns SVD file path ($info)', async ({ pname, expectedSvdPaths }) => {
            await cbuildRunReader.parse(TEST_CBUILD_RUN_FILE);
            const svdFilePaths = cbuildRunReader.getSvdFilePaths('/my/pack/root', pname);
            expect(svdFilePaths.length).toEqual(svdFilePaths.length);
            for (let i = 0; i < svdFilePaths.length; i++) {
                // eslint-disable-next-line security/detect-object-injection
                expect(expectedSvdPaths[i]).toEqual(svdFilePaths[i]);
            }
        });

        it('returns empty SVD file path list if nothing is parsed', () => {
            const svdFilePaths = cbuildRunReader.getSvdFilePaths('/my/pack/root');
            expect(svdFilePaths.length).toEqual(0);
        });
    });
});
