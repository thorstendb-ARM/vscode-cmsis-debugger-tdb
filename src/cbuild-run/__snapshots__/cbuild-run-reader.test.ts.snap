// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`CbuildRunReader Parser successfully parses a *.cbuild-run.yml file 1`] = `
{
  "compiler": "AC6",
  "debug-sequences": [
    {
      "blocks": [
        {
          "execute": "Write32(0x20000000, 0xA5A5A5A5);
",
        },
        {
          "blocks": [
            {
              "execute": "Read32(0x20000004);
",
            },
          ],
          "if": "condition == true",
        },
        {
          "execute": "Write32(0x20000008, 0x5);
",
        },
      ],
      "name": "MySequence",
      "pname": "Core0",
    },
    {
      "blocks": [
        {
          "execute": "DAP_Delay(500000);  // Delay for 500ms
",
        },
      ],
      "name": "DebugDeviceUnlock",
      "pname": "Core0",
    },
  ],
  "debug-topology": {
    "debugports": [
      {
        "accessports": [
          {
            "address": 0,
            "apid": 0,
          },
          {
            "address": 1048576,
            "apid": 1,
          },
        ],
        "dpid": 0,
      },
    ],
    "processors": [
      {
        "apid": 0,
        "pname": "Core0",
      },
      {
        "apid": 1,
        "pname": "Core1",
      },
    ],
  },
  "debugger": {
    "clock": 10000000,
    "gdbserver": [
      {
        "pname": "Core0",
        "port": 3336,
      },
      {
        "pname": "Core1",
        "port": 3339,
      },
    ],
    "name": "<default>",
    "port": "swd",
    "start-pname": "Core0",
    "terminal": 4444,
  },
  "device": "MyVendor::MyDevice",
  "device-pack": "MyVendor::MyDevice@1.0.0",
  "generated-by": "csolution version 2.8.0",
  "output": [
    {
      "file": "out/MyApp.bin",
      "info": "generate by MyApp",
      "load": "image",
      "type": "bin",
    },
    {
      "file": "out/out/MyApp.axf",
      "info": "generate by MyApp",
      "load": "symbols",
      "type": "elf",
    },
  ],
  "programming": [
    {
      "algorithm": "\${CMSIS_PACK_ROOT}/MyVendor/MyDevice/1.0.0/Flash/algorithms/MyAlgorithm_Core0.FLM",
      "pname": "Core0",
      "ram-size": 131072,
      "ram-start": 536870912,
      "size": 524288,
      "start": 0,
    },
    {
      "algorithm": "\${CMSIS_PACK_ROOT}/MyVendor/MyDevice/1.0.0/Flash/algorithms/MyAlgorithm_Extern.FLM",
      "pname": "Core0",
      "ram-size": 131072,
      "ram-start": 536870912,
      "size": 33554432,
      "start": 3221225472,
    },
  ],
  "solution": "my-test.csolution.yml",
  "system-descriptions": [
    {
      "file": "\${CMSIS_PACK_ROOT}/MyVendor/MyDevice/1.0.0/Debug/SVD/MyDevice_Core0.svd",
      "pname": "Core0",
      "type": "svd",
    },
    {
      "file": "\${CMSIS_PACK_ROOT}/MyVendor/MyDevice/1.0.0/Debug/SVD/MyDevice_Core1.svd",
      "pname": "Core1",
      "type": "svd",
    },
  ],
  "system-resources": {
    "memory": [
      {
        "access": "rx",
        "from-pack": "MyVendor::MyDevice@1.0.0",
        "name": "Flash",
        "pname": "Core0",
        "size": 524288,
        "start": 0,
      },
      {
        "access": "rwx",
        "from-pack": "MyVendor::MyDevice@1.0.0",
        "name": "SRAM0",
        "pname": "Core0",
        "size": 4194304,
        "start": 33554432,
      },
      {
        "access": "rwx",
        "from-pack": "MyVendor::MyDevice@1.0.0",
        "name": "SRAM1",
        "pname": "Core0",
        "size": 2621440,
        "start": 134217728,
      },
    ],
  },
  "target-type": "My-Test-Target-Type",
}
`;
